beacon-chain/state/stateV0/getters_validator.go

```diff
@@ -116,13 +116,26 @@ func (b *BeaconState) ValidatorAtIndexReadOnly(idx types.ValidatorIndex) (iface.
 
 // ValidatorIndexByPubkey returns a given validator by its 48-byte public key.
 func (b *BeaconState) ValidatorIndexByPubkey(key [48]byte) (types.ValidatorIndex, bool) {
-	if b == nil || b.valMapHandler == nil || b.valMapHandler.IsNil() {
+	if b.hasInnerState() || b.state.Validators == nil {
 		return 0, false
 	}
 	b.lock.RLock()
 	defer b.lock.RUnlock()
-	idx, ok := b.valMapHandler.Get(key)
-	return idx, ok
+
+	numOfVals := len(b.state.Validators)
+	// Exit early if there are no validators.
+	if numOfVals == 0 {
+		return types.ValidatorIndex(0), false
+
+	}
+	// Search backwards, starting from the latest added
+	// validator.
+	for i := numOfVals - 1; i >= 0; i-- {
```

Comment:
This performance trade off seems off. If the set is getting so big that we can't maintain a map in memory, won't O(n) be really poor performance?